src/eth_rpc/servers/alchemy_rpc.rs

```diff
@@ -1,41 +1,83 @@
+    /// The provider for interacting with the Ethereum network.
     eth_provider: P,
 }
 
 impl<P: EthereumProvider> AlchemyRpc<P> {
+    /// Creates a new instance of [`AlchemyRpc`].
     pub const fn new(eth_provider: P) -> Self {
         Self { eth_provider }
     }
 }
 
 #[async_trait]
 impl<P: EthereumProvider + Send + Sync + 'static> AlchemyApiServer for AlchemyRpc<P> {
+    /// Retrieves the token balances for a given address.
     #[tracing::instrument(skip_all, ret, fields(address = %address, token_addresses = ?token_addresses))]
     async fn token_balances(&self, address: Address, token_addresses: Vec<Address>) -> RpcResult<TokenBalances> {
         tracing::info!("Serving alchemy_getTokenBalances");
 
+        // Set the block ID to the latest block
         let block_id = BlockId::Number(BlockNumberOrTag::Latest);
-        let handles = token_addresses.into_iter().map(|token_addr| {
-            let token = EthereumErc20::new(token_addr, &self.eth_provider);
-            let balance = token.balance_of(address, block_id);
 
-            TokenBalanceFuture::new(Box::pin(balance), token_addr)
-        });
+        Ok(TokenBalances {
+            address,
+            token_balances: join_all(token_addresses.into_iter().map(|token_address| async move {
+                // Create a new instance of `EthereumErc20` for each token address
+                let token = EthereumErc20::new(token_address, &self.eth_provider);
+                // Retrieve the balance for the given address
+                let token_balance = token.balance_of(address, block_id).await?;
+                Ok(TokenBalance { token_address, token_balance })
+            }))
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>, EthApiError>>()?,
+        })
+    }
+
+    /// Retrieves the metadata for a given token.
+    #[tracing::instrument(skip_all, ret, fields(token_address = %token_address))]
+    async fn token_metadata(&self, token_address: Address) -> RpcResult<TokenMetadata> {
+        tracing::info!("Serving alchemy_getTokenMetadata");
+
+        // Set the block ID to the latest block
+        let block_id = BlockId::Number(BlockNumberOrTag::Latest);
+        // Create a new instance of `EthereumErc20`
+        let token = EthereumErc20::new(token_address, &self.eth_provider);
+
+        // Await all futures concurrently to retrieve decimals, name, and symbol
+        let (decimals, name, symbol) =
+            futures::try_join!(token.decimals(block_id), token.name(block_id), token.symbol(block_id))?;
 
-        let token_balances = join_all(handles).await.into_iter().collect::<Result<Vec<_>, EthApiError>>()?;
+        // Return the metadata
+        Ok(TokenMetadata { decimals, name, symbol })
+    }
+
+    /// Retrieves the allowance of a given owner for a spender.
+    #[tracing::instrument(skip_all, ret, fields(token_address = %token_address, owner = %owner, spender = %spender))]
```

Comment:
I think we might be using the macro wrong. As far as I understand it, the `fields` can be used to display extra information (e.g. [`#[instrument(skip(self), fields(self.name = %self.name))]`](https://docs.rs/tracing/latest/tracing/attr.instrument.html#examples-1)). If we want to display the arguments to the function call, I think we can just not skip all. Also maybe here it might be worth to add `err`?

Reply:
Yes this is true just fixed, this triggered a clippy related to https://github.com/rust-lang/rust-clippy/issues/12281

will be fixed soon in one of the next release of rust after https://github.com/rust-lang/rust-clippy/pull/12805 is integrated. Before that I put an allow at the top of the file as a tmp patch